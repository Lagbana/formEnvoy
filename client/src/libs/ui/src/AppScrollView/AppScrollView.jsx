/* Scroll view with gradient edges
=============================== */

import React, { forwardRef, PropsWithChildren } from 'react';
import {
  View,
  StyleSheet,
  Animated,
  ViewStyle,
  ScrollViewProps,
  ScrollView,
} from 'react-native';
import { LinearGradient } from 'expo-linear-gradient';
import hexToRgba from 'hex-to-rgba';
import { Colors } from 'libs/ui/src/WebButton/node_modules/libs/ui/src/WebLoading/node_modules/colibri/styles';

export interface AppScrollViewProps extends ScrollViewProps {
  vertical?: boolean;
  horizontal?: boolean;
  start?: boolean;
  end?: boolean;
  size?: number;
  color?: string;
  scrollVar?: Animated.Value;
  style?: ViewStyle;
  contentContainerStyle?: ViewStyle;
}

export const AppScrollView = forwardRef<
  ScrollView,
  PropsWithChildren<AppScrollViewProps>
>((props, scrollViewRef) => {
  const {
    children,
    vertical = true,
    horizontal = false,
    start = true,
    end = true,
    size = 15,
    color = 'white',
    scrollVar = undefined,
    contentContainerStyle = undefined,
    style = undefined,
    ...scrollViewProps
  } = props;

  const scrollHandler = scrollVar
    ? Animated.event([{ nativeEvent: { contentOffset: { y: scrollVar } } }], {
        useNativeDriver: true,
      })
    : () => null;

  // styles

  const styles = StyleSheet.create({
    container: {
      flex: 1,
    },
    scrollview: {},
    topGradient: {
      height: size,
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
    },
    bottomGradient: {
      height: size,
      position: 'absolute',
      bottom: 0,
      left: 0,
      right: 0,
    },
    leftGradient: {
      width: size,
      position: 'absolute',
      top: 0,
      left: 0,
      bottom: 0,
    },
    rightGradient: {
      width: size,
      position: 'absolute',
      top: 0,
      right: 0,
      bottom: 0,
    },
  });

  return (
    <View style={[styles.container, style]}>
      <Animated.ScrollView
        {...scrollViewProps}
        ref={scrollViewRef || undefined}
        /* vertical={vertical} // scrollview doesn't take vertical, not sure why this was specified */ horizontal={
          horizontal
        }
        style={styles.scrollview}
        contentContainerStyle={{
          paddingVertical: size * 0.5,
          flexGrow: 1,
          ...contentContainerStyle,
        }}
        scrollEventThrottle={1}
        onScroll={scrollHandler}
      >
        {children}
      </Animated.ScrollView>
      {vertical && start && (
        <LinearGradient
          colors={[hexToRgba(Colors[color], 1), hexToRgba(Colors[color], 0)]}
          style={styles.topGradient}
        />
      )}
      {vertical && end && (
        <LinearGradient
          colors={[hexToRgba(Colors[color], 0), hexToRgba(Colors[color], 1)]}
          style={styles.bottomGradient}
        />
      )}
      {horizontal && start && (
        <LinearGradient
          colors={[hexToRgba(Colors[color], 1), hexToRgba(Colors[color], 0)]}
          start={{ x: 0, y: 0 }}
          end={{ x: 1, y: 0 }}
          style={styles.leftGradient}
        />
      )}
      {horizontal && end && (
        <LinearGradient
          colors={[hexToRgba(Colors[color], 0), hexToRgba(Colors[color], 1)]}
          start={{ x: 0, y: 0 }}
          end={{ x: 1, y: 0 }}
          style={styles.rightGradient}
        />
      )}
    </View>
  );
});
