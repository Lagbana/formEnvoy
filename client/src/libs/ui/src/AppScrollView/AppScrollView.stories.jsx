import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { AppScrollView } from './AppScrollView'

storiesOf('AppScrollView', module)
  .add('Standard type', () => (
    <AppScrollView
			children={'Hello App ScrollView'}
    />
  ))