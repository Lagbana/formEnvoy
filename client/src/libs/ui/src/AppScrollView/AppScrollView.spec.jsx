import React from 'react'
import { shallow, configure } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import { AppScrollView } from './AppScrollView'

configure({ adapter: new Adapter() })

test('<AppScrollView /> renders correctly', () => {
  const component = shallow(<AppScrollView />)
  component.unmount()
  expect(toJson(component)).toMatchSnapshot()
})
