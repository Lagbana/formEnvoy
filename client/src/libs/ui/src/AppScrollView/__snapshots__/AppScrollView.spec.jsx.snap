// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppScrollView /> renders correctly 1`] = `""`;
